rfcs/0131-sri-hashes.md

```diff
@@ -0,0 +1,48 @@
+---
+feature: sri-hashes
+start-date: 2022-07-25
+author: Winter
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Use SRI hashes and the `hash` parameter for fetchers.
+
+# Motivation
+[motivation]: #motivation
+
+Currently in Nixpkgs, the hash formats that are used in parameters to fetchers range from SRI and base-32 (the most common ones) to hex. In addition to this, packages use the algorithm-specific arguments (`sha256`, `sha512`, etc.) or the generic `hash` argument at random. This creates inconsistencies that we'd rather not have.
```

Comment:
What is SRI in this context?

Reply:
[Subresource Integrity](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity), hashes look like `sha256-2mE26ES+fYSWdfMr8uTsX2VVGTNMDQ9MXEk5E/L95UI=`. Maybe I should add at least a definition of SRI in the RFC?

Reply:
> This creates inconsistencies that we'd rather not have.

Why? This is not a motivation, just a claim.

Using the different hash encodings genuinely has value, for example I frequently just copy over the base16 hash from a `SHA256SUMS` file provided by upstream, requiring less button presses, round trips and most importantly allows others to quickly and easily match the hashes textually when reviewing.

Reply:
I think a link to what SRI is in the RFC would be good to have. I also didn't know what SRI was.

Reply:
> Using the different hash encodings genuinely has value...

I agree with this point, but at the same time most people use either the SRI hashes given by Nix's hash mismatch errors, or convert those to base-32.

Reply:
And this is the nice thing about the current system: You can just put in whatever your tool spits out into a fetcher and it'll be able to fetch and verify the result – it's not like hashes contain any information that's more comprehensible for humans in one format or the other…

Reply:
Sure, but most packages use whatever Nix gives them, which also leads to things like a lot of redundant usage of the `sha256` argument (when `hash` can/should be used):
```console
$ rg --files-with-matches "sha256 = \"sha256\-" | wc -l
    7388
```

Reply:
I don't see a problem with that, `hash = "sha256-…` is hardly better than `sha256 = "sha256-…`.

Reply:
"Less button presses"?
In Emacs it is just `shell-command` with `nix hash to-sri sha256:<...>` followed with `set-mark-command` and `kill-ring-save` the SRI output. With completions and everything else to help.
Seriously, less button presses is hardly a serious reason when we have shell and editor completions all around today.

And an unified name like `hash` is better, because, well, new and better hash technologies appear everywhere. E.g. NetBSD pkgsrc now uses BLAKE hashes as default.
If we choose to support it, or any other hash algorithm for that matter, it would be way easier to reuse `hash = blake-XXXXX...` instead of creating a new `blake = XXXXX...`.

Reply:
>Why? This is not a motivation, just a claim.

One of the reasons might be sha* are semantically mutually exclusive, but not syntactically – you can specify both sha256 and sha512 at the same time by accident (pretty easy when using an override), but you cannot specify multiple different hash attributes.

Reply:
@vojta001 That's why these interfaces tend to have an `assert` or two behind the scenes…

> `nix hash to-sri`

Well first of all, this is actually a very good point: There is *no stable interface* for converting hashes into the SRI format, because `nix-hash(1)` only supports SRI as an input, not an output format.

Reply:
> Using the different hash encodings genuinely has value, for example I frequently just copy over the base16 hash from a `SHA256SUMS` file provided by upstream, requiring less button presses, round trips and most importantly allows others to quickly and easily match the hashes textually when reviewing.

no need to convert the hash to base64

SRI hashes only require a prefix:

```nix
{
  sha256 = "1cc9110705165dc09dd09974dd7c0b6709c9351d6b6b1cef5a711055f891dd0f";
  sha256 = "sha256-1cc9110705165dc09dd09974dd7c0b6709c9351d6b6b1cef5a711055f891dd0f";
  hash = "sha256-1cc9110705165dc09dd09974dd7c0b6709c9351d6b6b1cef5a711055f891dd0f";
}
```

base16 hashes are found with
`grep -r -E 'sha256 = "[0-9a-f]{64}";'`

Reply:
> sha256 is de-facto standard in nixpkgs,

MD5 was de facto standard no more than a decade ago.

Also, this was not about "good enough algorithms", but about a better format for the expression - namely, `hash = "..."` instead of `hashAlgorithm = "..."`.

Reply:
> This creates inconsistencies that we'd rather not have.

@winterqt could you add something more specific here, considering this is the whole motivation for the RFC?

The discussion above has brought up some more points:

- semantic vs syntactical mutual exclusion
- unique attribute name rather than multiple, algo-specific ones
- ...

Reply:
@milahu: conversion would be needed because SRI is _currently_ only accepting base64.  However, it would be relatively easy to change: https://github.com/NixOS/nixpkgs/pull/199660